lib/cfndsl/orchestration_template.rb

```diff
@@ -161,182 +167,120 @@ def initialize(description = nil, &block)
+      container = instance_variable_get("@#{container_name}s")
+      return [] unless container
+
+      invalids = []
+      referred_by = RefHash.new { |h, k| h[k] = [] }
+      self_check = source_containers.first.eql?(container)
+
+      container.each_pair do |name, entry|
+        name = name.to_s
+        begin
+          refs = entry.build_references([], self_check && name, method)
+          refs.each { |r| referred_by[r.to_s] << name }
+        rescue RefCheck::SelfReference, RefCheck::NullReference => e
+          # Topological sort will not detect self or null references
+          invalids.push("#{container_name} #{e.message}")
         end
       end
-    end
-    # rubocop:enable Metrics/PerceivedComplexity
 
-    # rubocop:disable Metrics/PerceivedComplexity
-    def validate_conditions(errors = [])
-      conditions = (@Conditions || {})
-      parameter_refs = (@Parameters || {}).keys
-
-      conditions.each_pair do |logical_id, condition|
-        path = "/Conditions/#{logical_id}"
-
-        condition.visit_json(path) do |sub_path, value|
-          if value.nil?
-            errors << "Null value at #{sub_path} is not permitted"
-          elsif !(value_refs = value.respond_to?(:refs) ? value.refs - GLOBAL_REFS.keys : []).empty?
-            invalid_refs = value_refs - parameter_refs
-            errors << "Invalid References : Path #{sub_path} refers to #{invalid_refs}" unless invalid_refs.empty?
-          end
+      referred_by.each_pair do |ref, names|
+        unless valid_ref?(ref, source_containers)
+          invalids.push "Invalid Reference: #{container_name}s #{names} refer to unknown #{method == :condition_refs ? 'Condition' : 'Reference'} #{ref}"
         end
       end
 
-      # TODO: condition functions Fn::Not, Fn::Or, Fn::And can refer to other conditions, and these cannot be cyclic either
-      errors
+      begin
+        referred_by.tsort if self_check && invalids.empty? # Check for cycles
+      rescue TSort::Cyclic => e
+        invalids.push "Cyclic references found in #{container_name}s #{referred_by} - #{e.message}"
+      end
+
+      invalids
     end
     # rubocop:enable Metrics/PerceivedComplexity
 
-    # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/AbcSize
-    def validate_outputs(errors = [])
-      outputs = (@Outputs || {})
-      conditions = (@Conditions || {})
-      parameters = (@Parameters || {})
-      resources = (@Resources || {})
+    def validate
+      errors = check_refs || []
+      raise CfnDsl::Error, "#{errors.size} errors in template\n#{errors.join("\n")}" unless errors.empty?
 
-      outputs.each_pair do |logical_id, output|
-        path = "/Outputs/#{logical_id}"
+      self
+    end
 
-        unless !output.condition || conditions.include?(output.condition)
-          errors << "Invalid Condition : Output #{logical_id} refers to missing Condition #{output.condition}"
+    def check_rule_refs
```

Comment:
Not sure this is ever called and it needs refactoring.

Reply:
This is new, so looks like a broken merge. Are you able to take a look at it please?

```
$ git blame lib/cfndsl/orchestration_template.rb | grep check_rule_refs
778a0e87 (Cam Maxwell    2019-09-05 17:03:11 +1000 135)       invalids = check_resource_refs + check_output_refs + check_rule_refs
778a0e87 (Cam Maxwell    2019-09-05 17:03:11 +1000 171)     def check_rule_refs```